appcelerator/titanium_mobile
# fix(android): Resuming with intent "FLAG_ACTIVITY_MULTIPLE_TASK" can hang the app

**JIRA:**
https://jira.appcelerator.org/browse/TIMOB-27271

**Summary:**
- Modified to effectively restart app when the following intent flags are set:
  * `FLAG_ACTIVITY_MULTIPLE_TASK`
  * `FLAG_ACTIVITY_NEW_DOCUMENT`
- Above flags make it impossible to resume existing root activity. Android OS will always create a new activity instance for matching intent when using these flags.
- Modified "resume" intent to use main/launcher flags.
  * We don't want to resume with `FLAG_ACTIVITY_CLEAR_TOP` or other destructive flags because they will destroy existing app's child activities.

**Test:**
1. Create a classic Titanium app.
2. Set project name to "IntentTest". _(This is `<name/>` in "tiapp.xml".)_
3. Set project's "Application Id" to "com.appc.intent.test". _(This is `<id/>` in "tiapp.xml".)_
4. Set up the "app.js" with the below code.
5. Build and run on Android.
6. Open the Mac "Terminal".
7. CD (change directory) to: `~/Library/Android/sdk/platform-tools`
8. In the terminal, enter the following...
`./adb shell am start -n com.appc.intent.test/.IntenttestActivity -a android.intent.action.VIEW -d https://www.appcelerator.com -f 0x08000000`
9. Verify that the app restarts.
10. Verify displayed "Launch Intent" has "data" value set to "https://www.appcelerator.com".
11. In the terminal, enter the following...
`./adb shell am start -n com.appc.intent.test/.IntenttestActivity -a android.intent.action.VIEW -d https://www.google.com`
12. Verify that "Launch Intent" label changes to "New Intent". _(App should not restart.)_
13. Verify "New Intent" has "data" value set to "https://www.google.com".
14. Tap on the Android "Home" button. _(Do **NOT** press Back button.)_
15. In the terminal, enter the following...
`./adb shell am start -n com.appc.intent.test/.IntenttestActivity -a android.intent.action.VIEW -d https://www.axway.com`
16. Verify that the app is resumed and in the foreground.
17. Verify displayed "New Intent" has "data" value set to "https://www.axway.com".

```javascript
Ti.Android.rootActivity.addEventListener("newintent", function(e) {
	label.text = "New Intent:\n" + JSON.stringify(e.intent, null, 4);
	Ti.API.info("@@@ newintent: " + JSON.stringify(e.intent));
});

var window = Ti.UI.createWindow();
var scrollView = Ti.UI.createScrollView({
	scrollType: "vertical",
	width: Ti.UI.FILL,
	height: Ti.UI.FILL,
});
var label = Ti.UI.createLabel({
	text: "Launch Intent:\n" + JSON.stringify(Ti.Android.rootActivity.intent, null, 4),
	width: Ti.UI.FILL,
	height: Ti.UI.SIZE,
});
scrollView.add(label);
window.add(scrollView);
window.open();
```
